### PR TITLE
[WebNN EP] Re-implement Unsqueeze, Squeeze, Flatten with WebNN's reshape

### PR DESCRIPTION
**Pull request by Honry**
WebNN will not provide `unsqueeze`, `squeeze`, `flatten2d` ops, as it can be easily implemented by reshape.

### CODE REVIEW

**Comment by Honry** [comment]
@fdwr, @guschmue, PTAL, thanks!

File: onnxruntime/core/providers/webnn/builders/impl/squeeze_unsqueeze_op_builder.cc [hunk 1]
```diff
@@ -78,15 +78,25 @@ Status SqueezeUnsqueezeOpBuilder::AddToModelBuilderImpl(ModelBuilder& model_buil
     }
   }
 
-  if (axes_data.size() > 0) {
-    options.set("axes", emscripten::val::array(axes_data));
-  }
-
   emscripten::val output = emscripten::val::undefined();
   if (op_type == "Squeeze") {
+    if (axes_data.size() > 0) {
```
Comment (fdwr) [inline onnxruntime/core/providers/webnn/builders/impl/squeeze_unsqueeze_op_builder.cc:83]: How about `!axes_data.empty()`?

File: onnxruntime/core/providers/webnn/builders/impl/squeeze_unsqueeze_op_builder.cc [hunk 2]
```diff
@@ -78,15 +78,25 @@ Status SqueezeUnsqueezeOpBuilder::AddToModelBuilderImpl(ModelBuilder& model_buil
...
+      options.set("axes", emscripten::val::array(axes_data));
+    }
     output = model_builder.GetBuilder().call<emscripten::val>("squeeze", input, options);
```
Comment (fdwr) [inline onnxruntime/core/providers/webnn/builders/impl/squeeze_unsqueeze_op_builder.cc:86]: I presume `squeeze` -> `reshape` will be handled in a later CR.

Comment (Honry) [reply onnxruntime/core/providers/webnn/builders/impl/squeeze_unsqueeze_op_builder.cc:86]: Yep, and also `flatten2d`. I should merge them, but another CR is fine.

File: onnxruntime/core/providers/webnn/builders/impl/squeeze_unsqueeze_op_builder.cc [hunk 3]
```diff
@@ -78,15 +78,25 @@ Status SqueezeUnsqueezeOpBuilder::AddToModelBuilderImpl(ModelBuilder& model_buil
...
   } else if (op_type == "Unsqueeze") {
-    output = model_builder.GetBuilder().call<emscripten::val>("unsqueeze", input, options);
+    // Use WebNN's reshape to implement Unsqueeze.
+    std::vector<int32_t> new_shape;
```
Comment (fdwr) [inline onnxruntime/core/providers/webnn/builders/impl/squeeze_unsqueeze_op_builder.cc:89]: `uint32_t`?

```
partial interface MLGraphBuilder {
  MLOperand reshape(MLOperand input, sequence<unsigned long?> newShape);
};
```

https://www.w3.org/TR/webnn/#api-mlgraphbuilder-reshape-method

File: onnxruntime/core/providers/webnn/builders/impl/squeeze_unsqueeze_op_builder.cc [hunk 4]
```diff
@@ -78,15 +78,25 @@ Status SqueezeUnsqueezeOpBuilder::AddToModelBuilderImpl(ModelBuilder& model_buil
...
+    std::transform(
+        input_shape.begin(), input_shape.end(), std::back_inserter(new_shape),
+        [](int64_t data) -> int32_t { return SafeInt<int32_t>(data); });
```
Comment (fdwr) [inline onnxruntime/core/providers/webnn/builders/impl/squeeze_unsqueeze_op_builder.cc:92]: This int64 to uint32 is repeated often. Consider adding a mini-helper function.

File: onnxruntime/core/providers/webnn/builders/impl/squeeze_unsqueeze_op_builder.cc [hunk 5]
```diff
@@ -78,15 +78,25 @@ Status SqueezeUnsqueezeOpBuilder::AddToModelBuilderImpl(ModelBuilder& model_buil
...
+    // Sort axes_data in ascending order.
+    std::sort(axes_data.begin(), axes_data.end());
+    // Expand new_shape according to axes_data.
+    for (const int32_t& axis : axes_data) {
+      new_shape.insert(new_shape.begin() + axis, 1);
```
Comment (fdwr) [inline onnxruntime/core/providers/webnn/builders/impl/squeeze_unsqueeze_op_builder.cc:97]: Is the `axis` is validated before reaching the WebNN EP?

Comment (fdwr) [inline onnxruntime/core/providers/webnn/builders/impl/squeeze_unsqueeze_op_builder.cc:97]: Do we handle negative axes?

In the DML EP, there is this line `HandleNegativeAxes`, which also validates the axis range:

```c++
    void UnsqueezeHelper::Initialize(
        IKernelInformationAdapter const& kernelInformation,
        IShapeInformationAdapter const& shapeInformation,
        uint32_t opsetVersion
        )
    {
        if (opsetVersion >= 13) // Axes are a dynamic input parameter.
        {
            ReadCpuLocalTensorIntoInt32(kernelInformation.GetConstantInputTensor(1), /*out*/ m_axes);
        }
        else // Axes were a constant attribute parameter.
        {
            m_axes = kernelInformation.GetAttributes().GetOptionalAttributeVectorInt32(AttrName::Axes);
        }
        std::vector<DimensionType> inputDimensions = shapeInformation.GetInputTensorShape(0);

        const uint32_t outputDimensionCount = gsl::narrow_cast<uint32_t>(inputDimensions.size() + m_axes.size());
        HandleNegativeAxes(/*inout*/ m_axes, outputDimensionCount);
        std::sort(m_axes.begin(), m_axes.end());
    }
```

**Review by fdwr — COMMENTED** [review]
Some comments. Thanks for accommodating the absence of `unsqueeze` in WebNN.

File: onnxruntime/core/providers/webnn/builders/impl/squeeze_unsqueeze_op_builder.cc (hunk 4 above)
Comment (Honry) [reply onnxruntime/core/providers/webnn/builders/impl/squeeze_unsqueeze_op_builder.cc:92]: This has been handled in [`HandleNegativeAxis`](https://github.com/microsoft/onnxruntime/blob/main/onnxruntime/core/providers/common.h#L30).

File: onnxruntime/core/providers/webnn/builders/impl/squeeze_unsqueeze_op_builder.cc (hunk 5 above)
Comment (Honry) [reply onnxruntime/core/providers/webnn/builders/impl/squeeze_unsqueeze_op_builder.cc:97]: Yes, handled in [`HandleNegativeAxis`](https://github.com/microsoft/onnxruntime/blob/main/onnxruntime/core/providers/common.h#L30), called at line 69 and 77 of this file.

Comment (Honry) [reply onnxruntime/core/providers/webnn/builders/impl/squeeze_unsqueeze_op_builder.cc:97]: Handled. See above comment.

**Comment by Honry** [comment]
@fdwr, I just fix the squeeze and flatten as well in this PR, please take another look, thanks!

File: onnxruntime/core/providers/webnn/builders/impl/squeeze_unsqueeze_op_builder.cc (hunk 5 above)
Comment (fdwr) [reply onnxruntime/core/providers/webnn/builders/impl/squeeze_unsqueeze_op_builder.cc:110]: Sorry, I didn't see that call inside the lambda.

**Review by fdwr — APPROVED** [review]

**Review by guschmue — APPROVED** [review]

**Pull request merged**